### DISCUSSION TITLE
Determine the real current with revisions enabled in subscription

### DISCUSSION THREAD

**Discussion by marcuslindblom** [ravendb/ravendb#15959] [Q&A]
In a subscription batch where revisions are enabled. How can I know if the current is the latest current? I need to know if the document is the last revision or if it is a historical one.

**Comment by ayende** [comment]
How t meaningful is it in the real world, because immediately after the batch, there is a new version, what do you do then?

You can do something like this:

```
from "Orders" (Revisions=true) as t
select {
    Previous: t.Previous,
    Current: t.Current,
    LatestChangeVector: getMetadata(load(id(t.Current)))["@change-vector"],

}
```
And then check if `Current`'s change vector match `LatestChangeVector`

**Reply by marcuslindblom** [reply]
Ok, I see. I think I misunderstood the hole concept. I use subscription to sync real-time data to the browser via a web socket connection. So my use case is that I want the last document and I need to know if it has been deleted.

So using a subscription without revisions makes it impossible to know if the document is deleted because we don't get a "signal" in the subscription when something is deleted. Adding revisions to the party means we don't get the last document.

Do I understand it correctly?

**Reply by ayende** [reply]
I'm not sure that I understand here.
In this model, wouldn't you have a _subscription per user_ ?

That can be _incredibly_ expensive.

Given that you are using a web socket, why not use the changes API, which is meant for this and will let you know about the latest going on?

**Reply by marcuslindblom** [reply]
Yes, one subscription per user. We discussed this topic a way back and I understood that is was not so expensive because the subscription is not active if they are not logged on.

The interface has a subset of the data stored in indexeddb and I need to make sure, in real-time that the data in the browser is up to date. If I log out, log in a couple of days later, I need to get the data that changed when I was logged off.

I don't think changes api can solve that for me?

**Reply by ayende** [reply]
The actual problem is when they *do*  log in.
If I log in after 3 weeks, I have to go through all the changes that happened in the meantime.
In that case, wouldn't it be better to just get the current snapshot of the values?

That said, I don't follow why you can't use the revisions in this case, you stream the values to the client and the final result is the current one.

**Reply by marcuslindblom** [reply]
I store the latest change vector on the client so when you log in three weeks later, using the same computer and browser you only get the changes you need.

I will do some tests with the changes api and see if that will work.

**Reply by marcuslindblom** [reply]
You said current snapshot of the values? How can I get that using changes api because that exactly what I need. Then I could use changes api for real-time updates.

**Reply by ayende** [reply]
Try imagining that you have ~100K changes in the time period between logins.
Right now, with subscriptions, you need to stream through all of them.

What happens when you have a _new_ user?

And for `Changes()`, you get the document id and need to load it separately.

**Reply by marcuslindblom** [reply]
As of know I create a subscription and set the change vector to LastDocument so I get a snapshot the the latest data. The subscription is somewhat unique to get data that the user has "access" to. So when a new user logs on she only gets the latest changes for all documents. If I used changes I need another way to get the first snapshot. Maybe a simple query could work for that

**Reply by marcuslindblom** [reply]
@ayende if I have a change vector, is it possible to query and only get documents changed after the provided change vector? Like a delta update or what we might call it. So if I a user has 100 documents and since last login 5 more has been created. The user only need the last 5 and the changes on the 100 he already have.

**Reply by ayende** [reply]
Subscription gives you everything _after_ the change vector.  It doesn't give you snapshot of data before-hand.
The way to handle your scenario is best handled using the `@last-modified`

The actual problem you have is not with 5 documents that you haven't sent, but what happens when you have 500K.

**Reply by marcuslindblom** [reply]
Ok, I see. I will do some tests and maybe I get back with some questions. Thanks

**Reply by marcuslindblom** [reply]
@ayende do you have an example where I can use `@last-modified` in a query?

**Reply by ayende** [reply]
I believe that this requires an index

**Reply by marcuslindblom** [reply]
Ah, ok
